core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/pythonUDF/PythonUDFOpExec.java

```diff
@@ -441,69 +351,230 @@ private static void readArrowStream(FlightClient client, String descriptorPath,
+            }
+            arrowFields.add(field);
+
+        }
+        return new org.apache.arrow.vector.types.pojo.Schema(arrowFields);
+    }
+
+    private static void safeDeleteTempFile(String fileName) {
         try {
-            byte[] resultBytes = client.doAction(new Action("compute")).next().getBody();
-            Map<String, String> result = mapper.readValue(resultBytes, Map.class);
-            if (result.get("status").equals("Fail")) {
-                String errorMessage = result.get("errorMessage");
-                throw new Exception(errorMessage);
+            deleteTempFile(fileName);
+        } catch (Exception innerException) {
+            innerException.printStackTrace();
+        }
+    }
+
+    private void connectToServer(Location flightServerURI) throws InterruptedException {
+        boolean connected = false;
+        int tryCount = 0;
+        while (!connected && tryCount < MAX_TRY_COUNT) {
+
+            try {
+                flightClient = FlightClient.builder(memoryAllocator, flightServerURI).build();
+                connected = new String(communicate(flightClient, "healthcheck"), StandardCharsets.UTF_8).equals("Flight Server is up and running!");
+                if (!connected) Thread.sleep(WAIT_TIME_MS);
+            } catch (FlightRuntimeException e) {
+                System.out.println("Flight Client:\tNot connected to the server in this try.");
+                flightClient.close();
+                tryCount++;
             }
-            readArrowStream(client, "fromPython", resultQueue);
-        } catch (Exception e) {
-            closeAndThrow(client, e);
+        }
+        if (tryCount == MAX_TRY_COUNT)
+            throw new RuntimeException("Exceeded try limit of " + MAX_TRY_COUNT + " when connecting to Flight Server!");
+    }
+
+    private void preparePythonScriptFile() throws IOException {
+        pythonScriptPath = isDynamic ?
+                getPythonResourcePath(String.valueOf(new Random().nextLong())) + ".py"
+                : getPythonResourcePath(pythonScriptPath);
+        if (isDynamic) {
+            // dynamic -> create a temp file and write the code into the file
+            writeTempPythonFile();
+        } else {
+            // static -> check if the script file exists
+            File scriptFile = new File(pythonScriptPath);
+            if (!scriptFile.exists()) throw new FileNotFoundException("Script file doest not exist!");
+        }
+    }
+
+    private Iterator<Tuple> processOneBatch(boolean inputExhausted) throws RuntimeException {
+        writeArrowStream(flightClient, inputTupleBuffer, globalInputSchema, "toPython", batchSize);
+        executeUDF(flightClient);
+        Queue<Tuple> outputTupleBuffer = new LinkedList<>();
+        readArrowStream(flightClient, "fromPython", outputTupleBuffer);
+
+        if (inputExhausted) {
+            Queue<Tuple> extraTupleBuffer = new LinkedList<>();
+            communicate(flightClient, "input_exhausted");
```

Comment:
better make this "input_exhausted" string a public static final variable, similar comments goes to other message types like "fromPython"

Reply:
will use enum to declare them.